### DISCUSSION TITLE
Logging

### DISCUSSION THREAD

**Discussion by rgstephens** [NVIDIA-NeMo/Guardrails#526] [Ideas]
I'm getting to know understand the state machine and the code base better and have a question about logging going forward. Following the events is helpful to understand why flows are working as expected and the logging really helps with this. But, I don't see currently any production structured logging.

The struct log project has some [good points](https://www.structlog.org/en/stable/logging-best-practices.html#pretty-printing-vs-structured-output) about the importance of structured logging and logging in production.

> Colorful and pretty printed log messages are nice during development when you locally run your code.
> However, in production you should emit structured output (like JSON) which is a lot easier to parse by log aggregators.

I started a [branch](https://github.com/rgstephens/NeMo-Guardrails/tree/feat/debug-logfile-structlog) to add an option to write to a logfile with structlog. I think a debug log would also need to be supported with the `server` option.

Is a PR along these lines something of interest to the team?

**Comment by drazvan** [comment]
Thanks for starting this, @rgstephens! This is definitely of interest. We should work out if this should be a purely debug feature (like the debug file feature you started), or to also have an option to log these in production to the standard output. We also need to check if there would be significant performance implications. Happy to explore this further with you.
